### DISCUSSION TITLE
JPA를  사용하는 이유

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#220] [Spring Data]
ORM 이란

**Comment by Kimprodp** [comment]
JPA는 Java 언어를 통해 ORM을 사용하기 위해 제공되는 인터페이스의 모음으로, JPA를 통해 개발자는 데이터베이스와의 상호작용을 객체 지향적 방식으로 수행할 수 있으며, 복잡한 SQL 쿼리를 작성하지 않고도 데이터를 효율적으로 관리할 수 있다.

**[JPA를 사용하는 이유]**

1. 생산성
	- JPA는 반복적인 CRUD SQL 작업을 간단한 API 호출로 처리할 수 있게 해줌
	- 이는 개발 시간을 단축시키고, 개발자가 비즈니스 로직에 더 집중할 수 있게 해줌

2. 유지보수성 향상
	- JPA를 사용하면, 엔티티 클래스만으로 데이터 모델을 표현할 수 있음
	- 이는 코드의 가독성을 높이고, 변경 사항이 발생했을 때 유지보수를 용이하게 함

3. 패러다임의 불일치 해결
	- Java에서는 상속관계가 존재하는데, 데이터베이스에서는 이러한 객체의 상속관계를 지원하지 않음(상속 기능을 지원하는 DB도 있지만 객체 상속과는 다름)
	- JPA는 연관된 객체를 사용하는 시점에 SQL을 전달할 수 있고, 같은 트랜잭션 내에서 조회할 때 동일성도 보장하기 때문에 다양한 패러다임의 불일치를 해결할 수 있음

4. 성능
	- 지연 로딩(Lazy Loading)과 캐시를 포함한 다양한 성능 최적화 기법을 지원하여, 애플리케이션과 데이터베이스 사이에서 성능 최적화 기회를 제공
	- 같은 트랜잭션안에서는 같은 엔티티를 반환하기 때문에 데이터 베이스와의 통신 횟수를 줄일 수 있음
	- 또한, 트랜잭션을 commit하기 전까지 메모리에 쌓고 한번에 SQL을 전송

**Comment by seunghye218** [comment]
데이터 액세스 기술을 사용하는 Spring 기반 애플리케이션을 더 쉽게 구축할 수 있습니다.

애플리케이션에 대한 데이터 액세스 계층을 구현하는 것은 상당히 번거로울 수 있습니다. 가장 간단한 쿼리를 실행하려면 너무 많은 상용구 코드를 작성해야 합니다. 페이지 매김, Auditing, 기타 자주 필요한 옵션을 추가하면 결국 길을 잃게 됩니다.

Spring Data JPA는 실제로 필요한 만큼의 노력으로 데이터 액세스 계층의 구현을 크게 개선하는 것을 목표로 합니다. 개발자는 다양한 기술을 사용하여 리포지토리 인터페이스를 작성하면 Spring이 자동으로 이를 연결해 줍니다. 심지어 사용자 정의 파인더를 사용하거나 예제를 통해 쿼리를 작성하면 Spring이 쿼리를 작성해줍니다.

1. 더 이상 DAO 구현이 필요하지 않습니다. 인터페이스를 확장함으로써 표준 DAO에서 사용할 수 있는 표준 데이터 액세스에 가장 적합한 CRUD 방법을 얻을 수 있습니다.
2. 사용자 정의 액세스 메서드 및 쿼리
    - 인터페이스에서 새로운 메소드를 정의하기만 하면 됩니다.
    - `@Query` 주석을 사용하여 JPQL 쿼리 제공
    - Spring Data의 고급 사양 및 Querydsl 지원을 사용
    - JPA Named 쿼리를 통해 사용자 정의 쿼리 사용
3. Automatic 사용자 정의 쿼리: 정의된 모든 메서드를 분석하고 메서드 이름에서 쿼리를 자동으로 생성 하려고 시도합니다
4. Transaction Configuration: 클래스 수준에서 읽기 전용 `@Transactional` 주석을 사용하고, 읽기 전용이 아닌 메서드에 대해 재정의됩니다.
